chapter_preliminaries/linear-algebra.ipynb

```diff
@@ -0,0 +1,1881 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 46,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public NDArray l2Norm(NDArray w){\n",
+    "    return ((w.pow(2)).sum()).sqrt();\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 47,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "ND: () cpu() float32\n",
+       "5.\n"
+      ]
+     },
+     "execution_count": 47,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "NDArray u = nd.create(new float[] {3,-4});\n",
+    "l2Norm(u)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 112
+   },
+   "source": [
+    "In deep learning, we work more often\n",
+    "with the squared $L_2$ norm.\n",
+    "You will also frequently encounter the $L_1$ *norm*,\n",
+    "which is expressed as the sum of the absolute values of the vector elements:\n",
+    "\n",
+    "$$\\|\\mathbf{x}\\|_1 = \\sum_{i=1}^n \\left|x_i \\right|.$$\n",
+    "\n",
+    "As compared with the $L_2$ norm,\n",
+    "it is less influenced by outliers.\n",
+    "To calculate the $L_1$ norm, we compose\n",
+    "the absolute value function with a sum over the elements.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 48,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "ND: () cpu() float32\n",
+       "7.\n"
+      ]
+     },
+     "execution_count": 48,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "u.abs().sum()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 116
+   },
+   "source": [
+    "Both the $L_2$ norm and the $L_1$ norm\n",
+    "are special cases of the more general $L_p$ *norm*:\n",
+    "\n",
+    "$$\\|\\mathbf{x}\\|_p = \\left(\\sum_{i=1}^n \\left|x_i \\right|^p \\right)^{1/p}.$$\n",
+    "\n",
+    "Analogous to $L_2$ norms of vectors,\n",
+    "the *Frobenius norm* of a matrix $\\mathbf{X} \\in \\mathbb{R}^{m \\times n}$\n",
+    "is the square root of the sum of the squares of the matrix elements:\n",
+    "\n",
+    "$$\\|\\mathbf{X}\\|_F = \\sqrt{\\sum_{i=1}^m \\sum_{j=1}^n x_{ij}^2}.$$\n",
+    "\n",
+    "The Frobenius norm satisfies all the properties of vector norms.\n",
+    "It behaves as if it were an $L_2$ norm of a matrix-shaped vector.\n",
+    "Invoking the following function will calculate the Frobenius norm of a matrix.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 49,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "ND: () cpu() float32\n",
+       "6.\n"
+      ]
+     },
+     "execution_count": 49,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "l2Norm(nd.ones(new Shape(4,9)))"
```

Comment:
I used same function previously created since based on this https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html it does the same thing.